gtsam/sfm/DsfTrackGenerator.h

```diff
@@ -0,0 +1,207 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file DsfTrackGenerator.h
+ * @date July 2022
+ * @author John Lambert
+ * @brief Identifies connected components in the keypoint matches graph.
+ */
+
+#pragma once
+#include <Eigen/Core>
+
+#include <gtsam/base/DSFMap.h>
+#include <gtsam/geometry/Point2.h>
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace gtsam {
+
+typedef DSFMap<IndexPair> DSFMapIndexPair;
+typedef Eigen::MatrixX2i CorrespondenceIndices; // N x 2 array
+
+//struct Keypoints;
+using KeypointCoordinates = Eigen::MatrixX2d;
+
+
+struct Keypoints {
+  KeypointCoordinates coordinates;
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
+};
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
+// Mapping from each image pair to (N,2) array representing indices of matching keypoints.
+using MatchIndicesMap = std::map<IndexPair, CorrespondenceIndices>;
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement {
+  size_t i;
+  gtsam::Point2 uv;
+
+  NamedSfmMeasurement(size_t i, gtsam::Point2 uv) : i(i), uv(uv) {}
+};
+
+
+/**
+ * @brief Track containing 2D measurements associated with a single 3D point.
+ * Note: Equivalent to gtsam.SfmTrack, but without the 3d measurement.
+ * This class holds data temporarily before 3D point is initialized.
+ */
+class SfmTrack2d {
+ private:
+  std::vector<NamedSfmMeasurement> measurements_;
+
+ public:
+  // Default constructor.
+  SfmTrack2d() = default;
+
+  // Constructor from measurements.
+  SfmTrack2d(std::vector<NamedSfmMeasurement> &measurements) : measurements_(measurements) {}
+
+  // Add a measurement to the track.
+  void addMeasurement(const NamedSfmMeasurement &m) {
+    measurements_.emplace_back(m);
+  }
+
+  /// The measurement at index `idx`
+  NamedSfmMeasurement measurement(size_t idx) const { return measurements_[idx]; }
+
+  // Return all measurements in the track.
+  std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
+
+  /// Total number of measurements in this track.
+  size_t numberMeasurements() const { return measurements_.size(); }
```

Comment:
nit: I think Frank started using nrMeasurements() in another class.. but up to you.

Reply:
I think we changed it to ‘numberMeasurements’ before merging in SfmTrack.h (https://github.com/borglab/gtsam/blob/develop/gtsam/sfm/SfmTrack.h#L71)

Reply:
oh i see, thanks!